pyo3-derive-backend/src/frompy.rs

```diff
@@ -0,0 +1,428 @@
+            Err(::pyo3::exceptions::PyTypeError::py_err(err_msg))
+        )
+    }
+
+    /// Build match for tuple struct variant.
+    fn build_tuple_variant(&self, var_ident: &Ident, len: usize) -> TokenStream {
+        let enum_ident = self.enum_ident;
+        let mut ext: Punctuated<Expr, syn::Token![,]> = Punctuated::new();
+        let mut fields: Punctuated<Ident, syn::Token![,]> = Punctuated::new();
+        let mut field_pats = PatTuple {
+            attrs: vec![],
+            paren_token: Paren::default(),
+            elems: Default::default(),
+        };
+        for i in 0..len {
+            ext.push(parse_quote!(slice[#i].extract()));
+            let ident = Ident::new(&format!("_field{}", i), Span::call_site());
+            field_pats.elems.push(parse_quote!(Ok(#ident)));
+            fields.push(ident);
+        }
+
+        quote!(
+            match <::pyo3::types::PyTuple as ::pyo3::conversion::PyTryFrom>::try_from(obj) {
+                Ok(s) => {
+                    if s.len() == #len {
+                        let slice = s.as_slice();
+                        if let (#field_pats) = (#ext) {
+                            return Ok(#enum_ident::#var_ident(#fields))
+                        }
+                    }
+                },
+                Err(_) => {}
+            }
+        )
+    }
+
+    /// Build match for transparent enum variants.
+    fn build_transparent_variant(
+        &self,
+        var_ident: &Ident,
+        field_ident: Option<&Ident>,
+    ) -> TokenStream {
+        let enum_ident = self.enum_ident;
+        if let Some(ident) = field_ident {
+            quote!(
+                if let Ok(#ident) = obj.extract() {
+                    return Ok(#enum_ident::#var_ident{#ident})
+                }
+            )
+        } else {
+            quote!(
+                if let Ok(inner) = obj.extract() {
+                    return Ok(#enum_ident::#var_ident(inner))
+                }
+            )
+        }
+    }
+
+    /// Build match for struct variant with named fields.
+    fn build_struct_variant(
+        &self,
+        tups: &[(&'a Ident, ExprCall)],
+        var_ident: &Ident,
+    ) -> TokenStream {
+        let enum_ident = self.enum_ident;
+        let mut field_pats = PatTuple {
+            attrs: vec![],
+            paren_token: Paren::default(),
+            elems: Default::default(),
+        };
+        let mut fields: Punctuated<Expr, syn::Token![,]> = Punctuated::new();
+        let mut ext: Punctuated<Expr, syn::Token![,]> = Punctuated::new();
+        for (ident, ext_fn) in tups {
+            field_pats.elems.push(parse_quote!(Ok(#ident)));
+            fields.push(parse_quote!(#ident));
+            ext.push(parse_quote!(obj.#ext_fn.and_then(|o| o.extract())));
+        }
+        quote!(if let #field_pats = #ext {
+            return Ok(#enum_ident::#var_ident{#fields});
+        })
```

Comment:
I think if you generate this as a closure, you might be able to re-use the non-variant case. This would cut down on the amount of macro, and also might lead to performance benefits, as with this current form it'll try to extract all the fields. Whereas the struct form below gives up as soon as any one fails.

Something vaguely like the below. 

```
let build_struct_variant = build_struct(...);
quote!( || -> pyo3::PyResult<_> { #build_struct_variant })
```

If you change all of the variant cases to be like this, the overall enum logic could probably become something like:

```
#(if let Ok(variant) = (#var_extracts)() { return variant; })*
```